the ``lazy_matrix_mul`` module
=========================

using ``lazy_matrix_mul``
--------------------

first import ``matrix_mul``:

	>>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul

Test: expected behavior:
    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[5, 6], [7, 8]])
    matrix([[19, 22],
            [43, 50]])

Test: bad matrixes * 2:
    >>> lazy_matrix_mul([[1, 2, 3], [3, 4, 5]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
    ...
    ValueError: shapes (2,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)

Test: 2 good matrix:
    >>> lazy_matrix_mul([[1, 2], [3, 4], [3, 4]], [[5, 6, 1], [7, 8, 2]])
    matrix([[19, 22,  5],
            [43, 50, 11],
            [43, 50, 11]])

Test: one empty matrix:
    >>> lazy_matrix_mul([], [[3]])
    Traceback (most recent call last):
    ...  
    ValueError: shapes (1,0) and (1,1) not aligned: 0 (dim 1) != 1 (dim 0)

Test: one empty matrix:
    >>> lazy_matrix_mul([[]], [[3]])
    Traceback (most recent call last):
    ...
    ValueError: shapes (1,0) and (1,1) not aligned: 0 (dim 1) != 1 (dim 0)


Test: string as m_a:
    >>> lazy_matrix_mul("impeach", [[9]])
    Traceback (most recent call last):
    ...
    TypeError: Scalar operands are not allowed, use '*' instead

Test: string as m_b:
    >>> lazy_matrix_mul([[8]], "trump")
    Traceback (most recent call last):
    ...
    TypeError: Scalar operands are not allowed, use '*' instead

Test: ints:
    >>> lazy_matrix_mul([1, 3, 4], [[9]])
    Traceback (most recent call last):
    ...
    TypeError: object of type 'int' has no len()

Test: string in list:
    >>> lazy_matrix_mul([["impeach"]], [[2, 4, 6]])
    Traceback (most recent call last):
    ...
    TypeError: invalid data type for einsum

Test: string in m_b:
    >>> lazy_matrix_mul([[2, 4, 6]], [["impeach"]])
    Traceback (most recent call last):
    ...
    TypeError: invalid data type for einsum

Test: bad rows:
    >>> lazy_matrix_mul([[1, 2], [3, 4, 5], [3, 4]], [[5, 6, 1], [7, 8, 2]])
    Traceback (most recent call last):
    ...
    ValueError: setting an array element with a sequence.

Test: m_b bad rows:
    >>> lazy_matrix_mul([[1, 2], [3, 4], [3, 4]], [[5, 6, 1], [7, 2]])
    Traceback (most recent call last):
    ...
    ValueError: setting an array element with a sequence.

Test:  missing 1 arg:
    >>> lazy_matrix_mul([[3]])
    Traceback (most recent call last):
    ...
    TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

Test: missing 2 args:
    >>> lazy_matrix_mul()
    Traceback (most recent call last):
    ...
    TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'

Test: mismatch:
    >>> lazy_matrix_mul([[5, 6, 10], [7, 8]], [[5, 6], [7, 8]])
    Traceback (most recent call last):
    ...
    ValueError: setting an array element with a sequence.

Test: tricky string element:
    >>> lazy_matrix_mul([[5, '6'], [7, 8]], [[5, 6], [7, 8]])
    Traceback (most recent call last):
    ...
    TypeError: invalid data type for einsum


Test: string arg:
    >>> lazy_matrix_mul("blacklivesmatter", [[9]])
    Traceback (most recent call last):
    ...
    TypeError: Scalar operands are not allowed, use '*' instead
